mockwebserver/src/main/java/okhttp3/mockwebserver/MockWebServer.kt

```diff
@@ -0,0 +1,1111 @@
+import okhttp3.mockwebserver.SocketPolicy.DISCONNECT_AFTER_REQUEST
+import okhttp3.mockwebserver.SocketPolicy.DISCONNECT_AT_END
+import okhttp3.mockwebserver.SocketPolicy.DISCONNECT_AT_START
+import okhttp3.mockwebserver.SocketPolicy.DISCONNECT_DURING_REQUEST_BODY
+import okhttp3.mockwebserver.SocketPolicy.DISCONNECT_DURING_RESPONSE_BODY
+import okhttp3.mockwebserver.SocketPolicy.EXPECT_CONTINUE
+import okhttp3.mockwebserver.SocketPolicy.FAIL_HANDSHAKE
+import okhttp3.mockwebserver.SocketPolicy.NO_RESPONSE
+import okhttp3.mockwebserver.SocketPolicy.RESET_STREAM_AT_START
+import okhttp3.mockwebserver.SocketPolicy.SHUTDOWN_INPUT_AT_END
+import okhttp3.mockwebserver.SocketPolicy.SHUTDOWN_OUTPUT_AT_END
+import okhttp3.mockwebserver.SocketPolicy.SHUTDOWN_SERVER_AFTER_RESPONSE
+import okhttp3.mockwebserver.SocketPolicy.STALL_SOCKET_AT_START
+import okhttp3.mockwebserver.SocketPolicy.UPGRADE_TO_SSL_AT_END
+import okhttp3.mockwebserver.internal.duplex.DuplexResponseBody
+import okio.Buffer
+import okio.BufferedSink
+import okio.BufferedSource
+import okio.ByteString.Companion.encodeUtf8
+import okio.Sink
+import okio.Timeout
+import okio.buffer
+import okio.sink
+import okio.source
+import org.junit.rules.ExternalResource
+import java.io.Closeable
+import java.io.IOException
+import java.net.InetAddress
+import java.net.InetSocketAddress
+import java.net.ProtocolException
+import java.net.Proxy
+import java.net.ServerSocket
+import java.net.Socket
+import java.net.SocketException
+import java.security.SecureRandom
+import java.security.cert.CertificateException
+import java.security.cert.X509Certificate
+import java.util.Collections
+import java.util.Locale
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.Executors
+import java.util.concurrent.LinkedBlockingQueue
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.logging.Level
+import java.util.logging.Logger
+import javax.net.ServerSocketFactory
+import javax.net.ssl.SSLContext
+import javax.net.ssl.SSLSocket
+import javax.net.ssl.SSLSocketFactory
+import javax.net.ssl.TrustManager
+import javax.net.ssl.X509TrustManager
+import kotlin.math.min
+
+/**
+ * A scriptable web server. Callers supply canned responses and the server replays them upon request
+ * in sequence.
+ */
+class MockWebServer : ExternalResource(), Closeable {
+  private val requestQueue = LinkedBlockingQueue<RecordedRequest>()
+  private val openClientSockets =
+      Collections.newSetFromMap(ConcurrentHashMap<Socket, Boolean>())
+  private val openConnections =
+      Collections.newSetFromMap(ConcurrentHashMap<Http2Connection, Boolean>())
+  private val requestCount = AtomicInteger()
+  private var bodyLimit = Long.MAX_VALUE
+  private var serverSocketFactory: ServerSocketFactory = ServerSocketFactory.getDefault()
```

Comment:
please don’t make this eager. It needs to be lazy because it could be expensive and we don’t want to do the work twice.

Reply:
Changing when things get initialized has behavior consequences. Please roll back those changes?

Reply:
Yep. .Fixed.